provider/equinix/environ.go

```diff
@@ -0,0 +1,784 @@
+
+	return e.Destroy(ctx)
+}
+
+func (e *environ) InstanceTypes(context.ProviderCallContext, constraints.Value) (instances.InstanceTypesWithCostMetadata, error) {
+	i := instances.InstanceTypesWithCostMetadata{}
+	instances, err := e.supportedInstanceTypes()
+	if err != nil {
+		return i, errors.Trace(err)
+	}
+
+	i.InstanceTypes = instances
+	return i, nil
+}
+
+func (e *environ) Instances(ctx context.ProviderCallContext, ids []instance.Id) ([]instances.Instance, error) {
+	toReturn := make([]instances.Instance, len(ids))
+	var missingInstanceCount int
+
+	tags := set.NewStrings("juju-model-uuid=" + e.Config().UUID())
+
+	for i, id := range ids {
+		d, resp, err := e.equinixClient.Devices.Get(string(id), nil)
+		if err != nil && resp != nil && resp.Request.Response.StatusCode == http.StatusNotFound {
+			logger.Warningf("instance %s not found", string(id))
+			missingInstanceCount = missingInstanceCount + 1
+			continue
+		} else if err != nil {
+			return nil, errors.Annotatef(err, "looking up device with ID %q", id)
+		}
+
+		deviceTags := set.NewStrings(d.Tags...)
+		if tags.Intersection(deviceTags).IsEmpty() {
+			missingInstanceCount++
+			continue
+		}
+		toReturn[i] = &equinixDevice{e, d}
+	}
+
+	if missingInstanceCount > 0 {
+		if missingInstanceCount == len(toReturn) {
+			return nil, environs.ErrNoInstances
+		}
+		return toReturn, environs.ErrPartialInstances
+
+	}
+	return toReturn, nil
+}
+
+func (e *environ) PrecheckInstance(ctx context.ProviderCallContext, args environs.PrecheckInstanceParams) error {
+	return nil
+}
+
+func (e *environ) PrepareForBootstrap(ctx environs.BootstrapContext, controllerName string) error {
+	e.name = controllerName
+	return nil
+}
+
+func (*environ) Provider() environs.EnvironProvider {
+	return &environProvider{}
+}
+
+func (e *environ) SetConfig(cfg *config.Config) error {
+	e.ecfgMutex.Lock()
+	defer e.ecfgMutex.Unlock()
+	ecfg, err := providerInstance.newConfig(cfg)
+	if err != nil {
+		return errors.Annotate(err, "invalid config change")
+	}
+	e.ecfg = ecfg
+	return nil
+}
+
+var configImmutableFields = []string{}
+var configFields = func() schema.Fields {
+	fs, _, err := configSchema.ValidationSchema()
+	if err != nil {
+		panic(err)
+	}
+	return fs
+}()
+var configSchema = environschema.Fields{}
+var configDefaults = schema.Defaults{}
+
+func newConfig(cfg, old *config.Config) (*environConfig, error) {
+	// Ensure that the provided config is valid.
+	if err := config.Validate(cfg, old); err != nil {
+		return nil, errors.Trace(err)
+	}
+	attrs, err := cfg.ValidateUnknownAttrs(configFields, configDefaults)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if old != nil {
+		// There's an old configuration. Validate it so that any
+		// default values are correctly coerced for when we check
+		// the old values later.
+		oldEcfg, err := newConfig(old, nil)
+		if err != nil {
+			return nil, errors.Annotatef(err, "invalid base config")
+		}
+		for _, attr := range configImmutableFields {
```

Comment:
Is this required, as `configImmutableFields` is empty?

Reply:
Probably not @achilleasa do you know?